### PR TITLE
Migration of the module account_credit_control for Odoo v10

### PR DESCRIPTION
**Pull request by sylvainvh**

### CODE REVIEW

**Comment by pedrobaeza** [comment]
We are already migrating account_followup module. Why not using it?

**Comment by sylvainvh** [comment]
[edit] my bad ! It's a pull request. I'll analyze this module

@pedrobaeza it's a good question ! where I can find this module ? I don't find this module in the repo account-financial-tools

**Comment by sylvainvh** [comment]
@pedrobaeza I don't understand the difference between the module account_credit_control and the module account_followup ? It seems to be the same. Right ?

**Comment by pedrobaeza** [comment]
That's the question. There isn't difference! The migration is hold until 9.0 migration is completed, but 10.0 migration will be straight-forward. Please check https://github.com/OCA/account-financial-tools/pull/450

**Comment by sylvainvh** [comment]
For the moment I'll finish my PR (I want green :p ). 
We will see after that if we merge this module in the module account_followup

**Comment by adrienpeiffer** [comment]
@sylvainvh I juste created a PR on your branch to add the setup file

File: account_credit_control/demo/credit_control_demo.xml
Comment (adrienpeiffer) [inline account_credit_control/demo/credit_control_demo.xml]: Are these data still useful ? With experience, it is not a good idea to define account.account as demo data as longer as the default chart of account will not be installed if accounts are created.

**Review by adrienpeiffer — CHANGES_REQUESTED** [review]
Are the tests performed? I don't see any trace in the logs file.

**Comment by fclementic2c** [comment]
@sylvainvh functional test ok
:+1: 
Good Job !

**Comment by sylvainvh** [comment]
Now I just need some review

**Comment by pedrobaeza** [comment]
@cubells please review and check differences with account_followup

File: account_credit_control/__manifest__.py
Comment (cubells) [inline account_credit_control/__manifest__.py]: Correct version to 10.0.0.1.0

File: account_credit_control/__manifest__.py
Comment (cubells) [inline account_credit_control/__manifest__.py]: It must be https

File: account_credit_control/data/data.xml
Comment (cubells) [inline account_credit_control/data/data.xml]: It must be: odoo noupdate=1

File: account_credit_control/__manifest__.py
Comment (cubells) [inline account_credit_control/__manifest__.py]: Remove empty keys in dictionary.

File: account_credit_control/models/account_account.py
Comment (cubells) [inline account_credit_control/models/account_account.py]: Change order: from odoo import fields, models

File: account_credit_control/models/account_invoice.py
Comment (cubells) [inline account_credit_control/models/account_invoice.py]: Change order: _, api, fields, models

File: account_credit_control/models/credit_control_line.py
Comment (cubells) [inline account_credit_control/models/credit_control_line.py]: Same as above

File: account_credit_control/models/credit_control_policy.py
Comment (cubells) [inline account_credit_control/models/credit_control_policy.py]: same as above

File: account_credit_control/models/credit_control_run.py
Comment (cubells) [inline account_credit_control/models/credit_control_run.py]: same as above

File: account_credit_control/views/credit_control_line.xml
Comment (cubells) [inline account_credit_control/views/credit_control_line.xml]: You must use font-awesome icons if necessary. Gtk icons are not already available in odoo v.10.

File: account_credit_control/views/credit_control_line.xml
Comment (cubells) [inline account_credit_control/views/credit_control_line.xml]: Add name selector if you can, as it is recommended at https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-10.0

File: account_credit_control/views/credit_control_line.xml
Comment (cubells) [inline account_credit_control/views/credit_control_line.xml]: Add name selector if you can, as it is recommended at https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-10.0

File: account_credit_control/wizard/credit_control_communication.py
Comment (cubells) [inline account_credit_control/wizard/credit_control_communication.py]: same as above

**Review by cubells — CHANGES_REQUESTED** [review]
Code review

**Comment by cubells** [comment]
@sylvainvh 

README file information is incorrect, outdated and incomplete.

Can you change README file in order to complete Usage and Configuration information about module?

And README file must follow this oca style: https://raw.githubusercontent.com/OCA/maintainer-tools/master/template/module/README.rst

File: account_credit_control/__manifest__.py
Comment (pedrobaeza) [reply account_credit_control/__manifest__.py]: Actually 10.0.1.0.0

File: account_credit_control/wizard/credit_control_policy_changer.py
Comment (cubells) [inline account_credit_control/wizard/credit_control_policy_changer.py]: It must be field reconciled instead of reconcile_id

File: account_credit_control/security/ir.model.access.csv
Comment (cubells) [inline account_credit_control/security/ir.model.access.csv]: Can you put fields in order? It's more readable. 

Perm fields later.

File: account_credit_control/migrations/0.3/post-migration.py
Comment (cubells) [inline account_credit_control/migrations/0.3/post-migration.py]: Post-migration script folder must be 10.0.1.x format not 0.3

**Review by cubells — CHANGES_REQUESTED** [review]
Functionally review

**Comment by cubells** [comment]
@sylvainvh :
It would be interesting to have some graphic or pivot view in order to analyze over due.

File: account_credit_control/migrations/0.3/post-migration.py
Comment (pedrobaeza) [reply account_credit_control/migrations/0.3/post-migration.py]: I bet that this migration script is for older versions and this should be removed.

File: account_credit_control/__manifest__.py
Comment (cubells) [reply account_credit_control/__manifest__.py]: It's true @pedrobaeza 

Thanks.

File: account_credit_control/models/account_invoice.py
Comment (guewen) [reply account_credit_control/models/account_invoice.py]: This is a non-blocking nitpicking, mind you

File: account_credit_control/models/credit_control_line.py
Comment (guewen) [inline account_credit_control/models/credit_control_line.py]: The correct error is `UserError` (same remark for the other files)

File: account_credit_control/models/account_invoice.py
Comment (pedrobaeza) [reply account_credit_control/models/account_invoice.py]: Indeed, but it's a nice to have if you don't mind doing the rest of the comments. If not, as said, don't block the PR for that.

File: account_credit_control/wizard/credit_control_policy_changer.py
Comment (sylvainvh) [reply account_credit_control/wizard/credit_control_policy_changer.py]: The field "reconcile" doesn't exist on the model account.move.line. I think it should be "full_reconcile_id". What do you think @cubells ?

Comment (cubells) [reply account_credit_control/wizard/credit_control_policy_changer.py]: Correct field, as I said above, is "reconciled". 

Field "full_reconcile_id" is a Many2one field for store reconcile_id when reconcile is full. This field is created by module "account_full_reconcile", that is not in dependencies.

**Comment by sylvainvh** [comment]
I made somes modification. Can you review ?

File: account_credit_control/wizard/credit_control_policy_changer.py
Comment (sylvainvh) [reply account_credit_control/wizard/credit_control_policy_changer.py]: Oké done on the last commit

File: account_credit_control/README.rst
Comment (cubells) [inline account_credit_control/README.rst]: It's "Invoicing" now, not "Accounting"

File: account_credit_control/README.rst
Comment (cubells) [inline account_credit_control/README.rst:26]: There isn't an Accounting tab in company form now. There is a "General information" tab,

Comment (cubells) [reply account_credit_control/README.rst:26]: Please add some information here in configuration about accounts tab in every policy because if user doesn't enter any account, policy per partner can't be selected.

File: account_credit_control/README.rst
Comment (cubells) [inline account_credit_control/README.rst]: "Invoicing" please
"Reports" please

File: account_credit_control/README.rst
Comment (cubells) [inline account_credit_control/README.rst:66]: You must add Try button:
`
.. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
   :alt: Try me on Runbot
   :target: https://runbot.odoo-community.org/runbot/92/10.0`

Comment (cubells) [reply account_credit_control/README.rst:66]: You must add "Bug Tracker" section and "Maintainer" section as is indicated in maintainers example:

https://raw.githubusercontent.com/OCA/account-financial-tools/10.0/account_check_deposit/README.rst

File: account_credit_control/__manifest__.py
Comment (cubells) [inline account_credit_control/__manifest__.py:7]: You must separate authors or you get a linter error:

Correct is  "Camptocamp, Odoo Community Association (OCA), Okia",

File: account_credit_control/README.rst
Comment (cubells) [inline account_credit_control/README.rst:3]: Please, change all headers of all files by this new format:

https://github.com/OCA/maintainer-tools/blob/master/template/module/__openerp__.py#L1-L3

**Review by cubells — CHANGES_REQUESTED** [review]

File: account_credit_control/README.rst
Comment (sylvainvh) [inline account_credit_control/README.rst]: I don't undersand, the name of the main menu is "Accounting". I don't see the menu "Invoicing"

File: account_credit_control/__manifest__.py
Comment (pedrobaeza) [reply account_credit_control/__manifest__.py:7]: No, that's not an actual problem

File: account_credit_control/README.rst
Comment (pedrobaeza) [reply account_credit_control/README.rst]: Maybe you should put both terms. If you only install module _account_, you have "Invoicing". When you install _account_accountant_, then the menu is renamed to "Accounting".

**Comment by jgrandguillaume** [comment]
Hi,

@cubells Thank you for your reviews, very much appreciated ! That being said, lot's of things here sounds like a bit picky for this PR if I may...

I've tested it, functionally speaking it it works great, Travis is green, so on my side :+1: to merge it.

Regards,

Joël

**Comment by pedrobaeza** [comment]
@jgrandguillaume although picky, we have imposed ourselves these high standards to conform 100% of the OCA guidelines. As we are making the effort to dispose the time spent on account_followup as discussed in https://github.com/OCA/account-financial-tools/pull/450 in favor of this module, and still needing by our part to:

* migrate to v9
* create migration scripts

this is not asking too much IMO.

File: account_credit_control/views/res_partner.xml
Comment (adrienpeiffer) [inline account_credit_control/views/res_partner.xml]: account_credit_control.group_account_credit_control_use**R**

File: account_credit_control/views/credit_control_line.xml
Comment (adrienpeiffer) [inline account_credit_control/views/credit_control_line.xml]: Add security groups here

File: account_credit_control/views/account_invoice.xml
Comment (adrienpeiffer) [inline account_credit_control/views/account_invoice.xml]: I don't really understand this condition. It means that the field is invisible if this one is empty. So, it will always invisible ?

**Review by adrienpeiffer — CHANGES_REQUESTED** [review]

File: account_credit_control/views/credit_control_run.xml
Comment (adrienpeiffer) [inline account_credit_control/views/credit_control_run.xml]: Change this one to fontawesome icon. e.g. fa-cog

**Review by adrienpeiffer — CHANGES_REQUESTED** [review]

**Comment by adrienpeiffer** [comment]
The PDF report isn't very well proportionate

acc2

File: account_credit_control/views/res_partner.xml
Comment (yvaucher) [reply account_credit_control/views/res_partner.xml]: good catch

**Comment by grindtildeath** [comment]
@sylvainvh 
I did some improvements according to review and will open a PR on your branch as soon as I will have fixed the PDF report.

**Comment by sylvainvh** [comment]
Hey, thank you for your feedback !!! I have to take the time to complete this PR.
@grindtildeath Thank you for you PR. I'll check that ASAP (it means now)

File: account_credit_control/views/account_invoice.xml
Comment (grindtildeath) [reply account_credit_control/views/account_invoice.xml]: The field is readonly, so if it isn't defined on the invoice there's no use to display it. I just removed the widget so it only appears as a link.

**Comment by grindtildeath** [comment]
Opened : https://github.com/QANSEE/account-financial-tools/pull/3

File: account_credit_control/views/account_invoice.xml
Comment (adrienpeiffer) [reply account_credit_control/views/account_invoice.xml]: @grindtildeath Ok, and why this one is readonly and when it is filled ?
cc\ @sylvainvh

File: account_credit_control/views/account_invoice.xml
Comment (adrienpeiffer) [inline account_credit_control/views/account_invoice.xml:22]: @grindtildeath why this one is readonly and when it is filled ?
cc\ @sylvainvh

**Comment by grindtildeath** [comment]
@adrienpeiffer While testing, I did fill it using the wizard 'credit.control.policy.changer'. 
Now I don't know if its purpose is to change an existing policy on an invoice or to define it the first time. Maybe @fclementic2c or @jgrandguillaume who did functional tests can let us know.

File: account_credit_control/views/account_invoice.xml
Comment (grindtildeath) [reply account_credit_control/views/account_invoice.xml:22]: @adrienpeiffer While testing, I did fill it using the wizard 'credit.control.policy.changer'.
Now I don't know if its purpose is to change an existing policy on an invoice or to define it the first time. Maybe @fclementic2c or @jgrandguillaume who did functional tests can let us know.

Comment (adrienpeiffer) [reply account_credit_control/views/account_invoice.xml:22]: Ok, I didn't catch the use of this wizard. So, it's ok for me :smile:

**Review by adrienpeiffer — APPROVED** [review]

**Comment by sylvainvh** [comment]
@grindtildeath I think there is a problem with the last PR. I'll try to see where is the problem

**Comment by sylvainvh** [comment]
Fixed :p (or not)

**Comment by grindtildeath** [comment]
@adrienpeiffer It seems there was an issue with my user record on OCA website. Should be fixed now, even if I don't know why it suddenly appears here after having already contributed to many PRs :smile:

**Comment by sylvainvh** [comment]
Youpiiii !!! Everything work. 
We can merge it now ? :p @pedrobaeza

**Comment by benwillig** [comment]
Functionnal test ok & code review :+1:

**Review by benwillig — APPROVED** [review]

**Review by cubells — APPROVED** [review]
:+1: 
LGTM

good job @sylvainvh

**Comment by pedrobaeza** [comment]
One last thing to do before merging: please squash commits by logical set/author.

**Comment by sylvainvh** [comment]
Here we are ! I hope it should be oke for you @pedrobaeza  :) 
I squashed all my commits by "features" (migration, improvement, unittest).

**Comment by pedrobaeza** [comment]
Congratulation for the hard work and let's welcome the credit control module in v10!

**Pull request merged**
